core/trino-main/src/main/java/io/trino/operator/scalar/BlockSet.java

```diff
@@ -0,0 +1,246 @@
+        this.maxFill = calculateMaxFill(hashCapacity);
+        this.hashMask = hashCapacity - 1;
+
+        blockPositionByHash = new int[hashCapacity];
+        Arrays.fill(blockPositionByHash, EMPTY_SLOT);
+
+        this.elementBlocks = new Block[expectedSize];
+        this.elementPositions = new int[expectedSize];
+
+        this.containsNullElement = false;
+    }
+
+    /**
+     * Does this set contain the value?
+     */
+    public boolean contains(Block block, int position)
+    {
+        requireNonNull(block, "block must not be null");
+        checkArgument(position >= 0, "position must be >= 0");
+
+        if (block.isNull(position)) {
+            return containsNullElement;
+        }
+        return blockPositionByHash[getHashPositionOfElement(block, position)] != EMPTY_SLOT;
+    }
+
+    /**
+     * Add the value to this set.
+     *
+     * @return {@code true} if the value was added, or {@code false} if it was
+     * already in this set.
+     */
+    public boolean add(Block block, int position)
+    {
+        requireNonNull(block, "block must not be null");
+        checkArgument(position >= 0, "position must be >= 0");
+
+        // containsNullElement flag is maintained so contains() method can have a shortcut for null value
+        if (block.isNull(position)) {
+            if (containsNullElement) {
+                return false;
+            }
+            containsNullElement = true;
+        }
+
+        int hashPosition = getHashPositionOfElement(block, position);
+        if (blockPositionByHash[hashPosition] == EMPTY_SLOT) {
+            addNewElement(hashPosition, block, position);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Returns the number of elements in this set.
+     */
+    public int size()
+    {
+        return size;
+    }
+
+    /**
+     * Return the position of the value within this set, or -1 if the value is not in this set.
+     */
+    public int positionOf(Block block, int position)
+    {
+        return blockPositionByHash[getHashPositionOfElement(block, position)];
+    }
+
+    /**
+     * Writes all values to the block builder checking the memory limit after each element is added.
+     */
+    public void getAllWithSizeLimit(BlockBuilder blockBuilder, String functionName, DataSize maxFunctionMemory)
+    {
+        long initialSize = blockBuilder.getSizeInBytes();
+        long maxBlockMemoryInBytes = toIntExact(maxFunctionMemory.toBytes());
+        for (int i = 0; i < size; i++) {
+            elementType.appendTo(elementBlocks[i], elementPositions[i], blockBuilder);
+            if (blockBuilder.getSizeInBytes() - initialSize > maxBlockMemoryInBytes) {
+                throw new TrinoException(
+                        EXCEEDED_FUNCTION_MEMORY_LIMIT,
+                        "The input to %s is too large. More than %s of memory is needed to hold the intermediate hash set.\n".formatted(functionName, maxFunctionMemory));
```

Comment:
Should this say "output hash set"? Seems like "intermediate" is a leftover from when `TypedSet` stored the set in memory itself.